### PR TITLE
[HTTP] Scavange fix

### LINKED ISSUE #61505: [HTTP] RemoveStalePools should never run in parallel
In certain cases, when there is a lot of different hosts and thus a lot of pools in pool manager (around 1 million from the customer report). And if scavenge timer (`_cleaningTimer`, `_cleanPoolTimeout`) gets triggered very often. The previous callback might still be running (even multiple of them) and thus causing rising number of handles.

The timer itself should never in parallel, we should start the cleanup logic while a previous callback is still processing.

### LINKED ISSUE #61506: [SslStream] Dispose seems to call blocking IO
> In particular that it looks like blocking I/O is happening as part of that cleanup.  That’s also something we should strive to avoid.
dispose

We should investigate and see if we can get rid of the blocking IO.

Discovered during investigation of the same problem as #61505

cc: @stephentoub @wfurt

### PR DESCRIPTION
**Pull request by ManickaP**
Disposes connection in a separate task to not to block the scavenging timer callback on the stream close.
Guards scavenging timer callback from parallel execution.

Fixes #61505
Fixes #61506

I plan to validate the fix with the customer before merging. But I'd like to know if the way I solved this is not completely wrong or too different from what you had in mind? @geoffkizer @stephentoub

### CODE REVIEW

**Comment by geoffkizer** [comment]
Rather than using a periodic timer and then checking for parallel execution, I would suggest we just use a single-shot timer and reset it in RemoveStalePools after the scavenge pass has completed.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 1]
```diff
@@ -1979,7 +1979,11 @@ public bool CleanCacheAndDisposeIfUnused()
             }
 
             // Dispose the stale connections outside the pool lock, to avoid holding the lock too long.
-            toDispose?.ForEach(c => c.Dispose());
+            // Dispose them asynchronously to not to block the caller on closing the SslStream or NetworkStream.
+            if (toDispose is not null)
+            {
+                Task.Run(() => toDispose.ForEach(c => c.Dispose()));
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1985]: This unnecessarily allocates a closure/delegate even if toDispose is null.  That can be avoided with:
```suggestion
                Task.Factory.StartNew(s => ((List<HttpConnectionBase>)s!).ForEach(c => c.Dispose()), toDispose,
                    CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
```

**Comment by geoffkizer** [comment]
One issue above, otherwise LGTM.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs [hunk 2]
```diff
@@ -479,25 +479,27 @@ private void RemoveStalePools()
         {
             Debug.Assert(_cleaningTimer != null);
 
-            // Iterate through each pool in the set of pools.  For each, ask it to clear out
-            // any unusable connections (e.g. those which have expired, those which have been closed, etc.)
-            // The pool may detect that it's empty and long unused, in which case it'll dispose of itself,
-            // such that any connections returned to the pool to be cached will be disposed of.  In such
-            // a case, we also remove the pool from the set of pools to avoid a leak.
-            foreach (KeyValuePair<HttpConnectionKey, HttpConnectionPool> entry in _pools)
+            try
```
[the remark this answers is not in the document]
Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs:482]: I'll remove it.
I didn't want to accidentally end up with turned off cleaning timer which never gets restarted.

**Review by geoffkizer — APPROVED** [review]

**Comment by ManickaP** [comment]
Customer validated the fix against 6.0 and 5.0. It helped with overlapping callbacks. However, in case of 5.0, they're still seeing abnormal number of `HttpConnectionPool` instances even after scavenging (this doesn't happen in 6.0 neither before nor after this change).

**Pull request merged**